src/models/ModelBuilder.cc

```diff
@@ -3,23 +3,136 @@
 //
 
 #include "ModelBuilder.h"
+#include <unordered_set>
 
-void ModelBuilder::processSubstitutions(Logic::SubstMap const & subst) {
-    Logic::SubstMap copy;
-    subst.copyTo(copy);
-    logic.substitutionsTransitiveClosure(copy);
-    auto assignCopy = this->assignment;
-    auto model = this->build();
-    for (auto key : copy.getKeys()) {
-        if (logic.isIte(key)) {
-            // We store only values of reals variables
+std::unique_ptr<Model> ModelBuilder::build() {
+    std::unordered_map<SymRef,TemplateFunction,SymRefHash> builtDefinitions;
+    for (auto & symbolSigVal : definitions) {
+        const ValuationNode * valuationNode = symbolSigVal.second.second;
+        SRef sr = logic.getSortRef(symbolSigVal.first);
+        PTRef body = valuationTreeToFunctionBody(valuationNode, sr);
+        TemplateFunction templateFun(std::move(symbolSigVal.second.first), body);
+        builtDefinitions[symbolSigVal.first] = std::move(templateFun);
+    }
+    return std::unique_ptr<Model>(new Model(logic, assignment, builtDefinitions));
+}
+
+void ModelBuilder::addToTheoryFunction(SymRef sr, vec<PTRef> vals, PTRef val)
+{
+    if (logic.getSortRef(sr) != logic.getSortRef(val)) {
+        throw OsmtApiException("Incompatible sort for symbol valuation");
+    }
+    Symbol & sym = logic.getSym(sr);
+    if (sym.nargs() != vals.size_()) {
+        throw OsmtApiException("Incorrect valuation for symbol: argument and valuation size do not match");
+    }
+    for (int i = 0; i < vals.size(); i++) {
+        if (sym[i] != logic.getSortRef(vals[i])) {
+            throw OsmtApiException("Incorrect valuation for symbol: sort mismatch");
+        }
+    }
+
+    if (not hasTheoryFunction(sr)) {
+        vec<PTRef> formalArgs; formalArgs.capacity(vals.size());
+        std::string symName(logic.getSymName(sr));
```

Comment:
This seams to be an unused variable? This should be caught by the compiler, we should check the CI.

Reply:
Yes, it's not caught for some reason.  Could it be because it uses the `std::string`'s `char*` constructor and could therefore have a side effect?

Reply:
That's possible, I don't know.